source/Halibut/Transport/ConnectionPool.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Runtime.Serialization;
+using Halibut.Diagnostics;
+using Halibut.Logging;
+
+namespace Halibut.Transport
+{
+    public class ConnectionPool<TKey, TPooledResource>
+        where TPooledResource : class, IPooledResource
+    {
+        readonly ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>> pool = new ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>>();
```

Comment:
What's the purpose of this change? It seems unrelated.

Reply:
It's to fix this malarkey. https://github.com/OctopusDeploy/Halibut/blob/master/source/Halibut/Transport/SecureClientConnectionPool.cs#L51-L81

The initial problem is a concurrency issue where the same object could be returned twice, leading to it being added to the pool twice. This fixes the concurrency issue and ensures the item is only added once. That's why there's a new `ShouldNotAllowMultipleReturnsOfSameConnection` test.

This fix was needed because the pool was not cleaned up properly as part of the dispose, which meant one connection was not disconnecting when it should have.